Step3. 리뷰 부탁드리겠습니다!

lkimilhol opened the pull request:
안녕하세요! 리뷰어님! 미션 제출이 조금 늦었습니다.

JPA를 실무에서 사용해본 경험이 없어서, 조금 공부를 하면서 질문 내용을 추리거나, 미션에 적용해 보는 등의 경험을 해보고 싶었는데, 막상 공부를 진행하니 JPA는 계속해서 제가 공부를 하면서 부족한 부분을 보완하고 미션 제출을 하는것이 시간상 좋을 것이라고 판단하여 제출합니다. ㅎㅎ

아래는 질문 내용인데요. 질문 내용이 조금 복잡 하실 수 있어 혹시라도 이해가 안되시면 코멘트 부탁드리겠습니다.

그러면 제가 dm으로 최대한 정리하여 다시 질문을 드리도록 하겠습니다.

마지막 리뷰 잘 부탁드리며 JPA 학습 관련하여 해주실 조언이 있다면 부탁드리겠습니다! (현재는 T아카데미, 영한님의 JPA 관련 강의 까지 들은 상태고 표준 ORM 책 정독하며 다시 내용을 복습하고 있습니다!)

감사합니다!


Q. 이번 미션의 목표는 deleteQuestion 메서드를 리팩토링 하는 부분이었는데요.  처음 구현한 방법은 Question 객체에서 Answers 라는 일급 컬렉션을 가지고 있으니 Answers 일급 컬렉션 안에서 Answer를 순회하며 Answer를 delete 하도록 하였습니다.  구현을 하고 테스트를 해보니 Mock에서 사용하지 않는 stubbing이 있었다는 에러 메시지와 함께 테스트가 실패하였습니다.  확인을 해보니 answerRepository의 findByQuestionIdAndDeletedFalse 메서드가 실행되지 않아 테스트가 실패하였는데요. 차라리 테스트 케이스의 내용을 좀 수정하고 일급 컬렉션 안에서 answer를 순회하며 delete를 하는것이 더 좋은 설계 였을까요?

현재 제출한 미션에서는 QnaService 내의 deleteQuestion 메서드를 deleteQna 메서드로 이름을 수정하고, question과 answer를 delete 해주는 로직을 분류 하였습니다. question과 answer 객체에서 각각 delete를 따로 구현하는 것을 생각하여 구현 하였는데요. 제가 개발한 방식도 괜찮은지 조금 의견을 들어보고 싶습니다!

lkimilhol committed:
refactor(test): 테스트 클래스 명 수정

- Repository 추가

lkimilhol committed:
refactor(DeleteHistory): contentType 타입 수정

- ordinary -> enumerated 수정

lkimilhol committed:
refactor(Answer): 연관관계 매핑 추가

lkimilhol committed:
refactor(DeleteHistory): 연관관계 매핑 추가

lkimilhol committed:
refactor(Question): 연관관계 매핑 추가

- DeleteHistory import 순서 변경

lkimilhol committed:
docs(README): 2단계 요구사항 추가

lkimilhol committed:
refactor(AnswerRepositoryTest,QuestionTest): 테스트케이스가 다른 테스트 케이스에 영향 받지 않도록 수정

lkimilhol committed:
refactor(Answer): setter 제거

lkimilhol committed:
refactor(Question): Answers 일급 컬렉션 생성

- List<Answer> -> 일급 컬렉션으로 수정
- 테스트 케이스 추가

lkimilhol committed:
refactor(Question,QuestionRepository,QnaService): import 문 정리

lkimilhol committed:
refactor(Question): 접근 제어자 누락 수정

lkimilhol committed:
test(AnswersTest): 글 답변의 삭제에 대한 권한 체크 로직 추가

lkimilhol committed:
docs(README): 3단계 요구사항 추가

lkimilhol committed:
test(QuestionRepositoryTest): 테스트 리팩토링

- 변수명을 한글로 하여 이해하기 쉽도록 수정
- 질문 테스트가 제대로 패스 되지 않는 이슈 해결
- 새로 생성하지 않아도 되는 변수 save 제거

lkimilhol committed:
refactor(Answer): 객체를 비교하도록 수정

- id와 비교를 잘못하는 부분 제거

lkimilhol committed:
refactor(Answer,Answers): Answers에서 답변을 삭제하는 부분 이동

- Answers 일급 컬렉션에서 delete 해주느 로직을 Answer로 이동

lkimilhol committed:
refactor(Question,QnaService): 글 삭제 로직 리팩토링

- QnaService에서 질문글과 답변을 체크하여 삭제 되도록 수정
- deleteQuestion을 deleteQna 메서드명 수정

lkimilhol committed:
refactor(Question,Answer): 에러 메시지 추출

- 메시지 내용을 상수로 수정

lkimilhol committed:
refactor(Question): 리팩토링

- import 순서 수정
- Stream.of 사용

lkimilhol committed:
docs(README): 질문 사항 추가

lkimilhol committed:
Merge branch 'lkimilhol' into step3

# Conflicts:
#	README.md
#	src/main/java/qna/domain/Answer.java
#	src/main/java/qna/domain/Question.java
#	src/main/java/qna/service/QnaService.java
#	src/test/java/qna/domain/QuestionTest.java

lkimilhol committed:
test(QuestionTest): import 수정

lkimilhol committed:
refactor(QnaService): Answer의 삭제를 Question을 통해서 하도록 수정

- Question은 Answers를 포함하고 있으므로 레퍼지토리에서 데이터를 가져오지 않도록 수정
- QnaService에서 mock 객체 테스트 에러를 수정

lkimilhol commented:
리뷰어님! 고민 끝에 Question의 Answers를 통하여 소프트 딜리트 되도록 수정하였습니다!
deleteQuestion 메서드는 줄어들었지만 결국 서비스 레이어 자체에서 삭제에 관한 로직이 그대로 남아있기 때문입니다!
참고 부탁드리겠습니다 :)

lkimilhol committed:
refactor(Answers): 이미 지워진 글에 대한 처리

- 이미 지워진 경우에는 다시 지울 필요가 없으므로 수정

lkimilhol committed:
refactor(QnaService,DeleteHistoryList): 일급 컬렉션 생성 리팩토링

- List<DeleteHistory> -> 일급 컬렉션 생성

hhlyn0328 commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -21,4 +23,18 @@ public int size() {
         answers.add(answer);
     }
+
+    public List<DeleteHistory> delete(User loginUser) throws CannotDeleteException {
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
+        for (Answer answer : answers) {
+            addDeleteHistory(loginUser, deleteHistories, answer);
+        }
+        return deleteHistories;
+    }
+
+    private void addDeleteHistory(User loginUser, List<DeleteHistory> deleteHistories, Answer answer) throws CannotDeleteException {
+        if (!answer.isDeleted()) {
+            deleteHistories.add(answer.delete(loginUser));
+        }
+    }
```
deleteHistories 를 던져서 다른 method 에서 수정하고 있는데요
이런 경우 변경지점을 추적하기 어려울 수 있습니다

stream . fileter .map 을 이용해볼 수 있지 않을까요?

lkimilhol replied:
stream을 사용하도록 하였는데, 스트림의 내용중 예외처리 부분이 너무 혼잡하여 CannotDeleteException을 RuntimeException을 상속 받도록 수정해보았습니다. 구글링을 통해 FuntinalInterface를 생성하여 한번 포장하는 기법도 있었는데요. 오버 엔지니어링이라고 생각이 되어서 RuntimeException 수정 방향으로 해보았는데요. 혹시라도 문제 될만한 부분이 있다면 피드백 부탁드리겠습니다!

hhlyn0328 replied:
JPA에서 Checked Exception / UnChecked Exception 은 Rollback 과도 연관되어있어서 중요한 부분인데요 
이 경우에, 중간에 삭제가 실패할 경우 다 실패되는게 맞지 않을까 싶어서요
RuntimeException 으로 수정한 방향에 대해 찬성입니다 :)

lkimilhol replied:
피드백 감사합니다! 이 부분도 공부해보겠습니다!

hhlyn0328 commented on src/main/java/qna/domain/DeleteHistoryList.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class DeleteHistoryList {
```
일급컬렉션의 이름은 복수형으로 선언해주면 좋을거 같아요 
내부의 자료구조가 List 에서 Set 으로 변경되면 
객체명도 따라서 변경해야하거나, 그대로 두면 혼선을 줄 수 있기 때문입니다!

lkimilhol replied:
네 맞습니다! 전에 자바지기님의 강의 중 질문 내용중 하나네요 ㅠ 수정했습니다!

hhlyn0328 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -112,4 +105,19 @@ public String toString() {
+        checkWriter(loginUser);
+        setDeleted(true);
+        DeleteHistoryList deleteHistories = new DeleteHistoryList();
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, id, getWriter(), LocalDateTime.now()));
+        deleteHistories.add(getAnswers().delete(loginUser));
```
writer와 answers 가 question의 instance 변수여서 getter를 쓰지 않아도 될 거 같아요!

lkimilhol replied:
맞네요! 수정하였습니다!

hhlyn0328 commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -35,25 +29,7 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistoryList deleteHistories = findQuestionById(questionId).delete(loginUser);
```
questionId에 해당하는 question이 없는 경우에 대한 처리도 추가되면 좋을 거 같아요!

lkimilhol replied:
findQuestionById 메서드를 통하여 findByIdAndDeletedFalse 메서드를 호출할때 orElseThrow 메서드를 통하여 처리가 될 것으로 예상을 하긴 했는데요! findById 메서드를 호출하여 이에 대한 유효성 검사를 추가 해보자라는 의미이실까요?!

hhlyn0328 replied:
죄송합니다 ㅠㅠ 
제가.. 변경내역을 중점으로 보다보니 query method라고 착각했네요 😢

lkimilhol replied:
아닙니다!! 피드백 주셔 감사합니다 :)

hhlyn0328 commented:
`마지막 리뷰 잘 부탁드리며 JPA 학습 관련하여 해주실 조언이 있다면 부탁드리겠습니다! (현재는 T아카데미, 영한님의 JPA 관련 강의 까지 들은 상태고 표준 ORM 책 정독하며 다시 내용을 복습하고 있습니다!)`

잘 진행하고 계신거 같아요 👍  영항님의 책과 강의가 많은 도움이 되실거에요 
JPA가 다음 미션들에서도 기본적으로 사용되기 때문에 지금 당장은 시간이 좀 걸려도 기초를 다지고 가시는게 좋습니다

JPA가 아무래도 데이터를 기반으로 하는 기술이다 보니 이론만 공부하는 것보다 
토이프로젝트를 개발해보시는게, 실제 쿼리가 날아가는 것도 볼 수 있고 
N+1 이슈, 고아객체와 같이 실무에서 성능이슈로 만날 수 있는 부분들을 직접 겪어볼 수 있어서 좋은거 같습니다

hhlyn0328 commented:
`Q. 이번 미션의 목표는 deleteQuestion 메서드를 리팩토링 하는 부분이었는데요. 처음 구현한 방법은 Question 객체에서 Answers 라는 일급 컬렉션을 가지고 있으니 Answers 일급 컬렉션 안에서 Answer를 순회하며 Answer를 delete 하도록 하였습니다. 구현을 하고 테스트를 해보니 Mock에서 사용하지 않는 stubbing이 있었다는 에러 메시지와 함께 테스트가 실패하였습니다. 확인을 해보니 answerRepository의 findByQuestionIdAndDeletedFalse 메서드가 실행되지 않아 테스트가 실패하였는데요. 차라리 테스트 케이스의 내용을 좀 수정하고 일급 컬렉션 안에서 answer를 순회하며 delete를 하는것이 더 좋은 설계 였을까요?`

이번 미션의 목표는 서비스의 있는 로직을 도메인에게 위임하는 것 인데요 
JPA 연관관계를 어떻게 맺고 있는지 cascade와 orphanRemoval 과 같은 설정에 따라서 내용은 달라질 수 있을 거 같아요
결국엔 repository와 의존해야하는 실제 조회하는 부분과 저장하는 부분만 service에 남아있으면 됩니다! 

그래서 지금 구현하신 방법도 좋다고 생각해요 ㅎㅎ

hhlyn0328 reviewed (changes requested):
질문 삭제하기 리팩터링을 잘 진행해주셨네요 👍 

질문에 대한 답변과 몇 가지 코멘트 추가했으니 참고해주세요
JPA 미션도 어느덧 끝이 보이네요 화이팅!!

lkimilhol committed:
refactor(Answers,CannotDeleteException): stream 을 사용하도록 리팩토링

- Answers의 delete 메서드를 stream을 사용하도록 수정
- Exception 처리의 메서드의 코드가 복잡성이 증가하여 CannotDeleteException을 RuntimeException을 상속 받도록 수정

lkimilhol committed:
refactor(DeleteHistoryList): 일급 컬렉션의 이름 수정

- List postfix를 제거

lkimilhol committed:
refactor(Question): getAnswers 메서드 제거

- 인스턴스에서 answers가 접근 가능 하여 getAnswers 제거

lkimilhol committed:
test(QnaServiceTest): 질문을 찾지 못한 경우 테스트 추가

- 질문 id에 해당하는 Question을 찾지 못하는 경우 테스트

hhlyn0328 reviewed (approved):
제가 저번에 잘못 코멘트 드렸던 부분이 있었는데도
리뷰내용 반영을 잘 해주셨네요 👍 

JPA 미션은 여기서 마무리하겠습니다 
끝까지 완주하시길 응원할게요 ! 🔥
